src/include/daos_api.h

```diff
@@ -220,11 +220,149 @@ daos_prop_alloc(uint32_t entries_nr);
+ * \param[in]	principal	Principal name, if type is USER or GROUP. NULL
+ *				otherwise.
+ *
+ * \return	Pointer to the matching ACE, or NULL if not found
+ */
+struct daos_ace *
+daos_acl_get_ace_for_principal(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal);
+
+/**
+ * Insert an Access Control Entry in the appropriate location in the ACE
+ * list. The expected order is: Owner, Users, Assigned Group, Groups, Everyone.
+ *
+ * This requires us to reconstruct and reallocate the ACL structure. Both the
+ * old structure and the new one must be freed by the caller. The old data is
+ * left unchanged.
+ *
+ * If the new ACE is an update of an existing entry, it will replace the old
+ * entry.
+ *
+ * \param[in]	acl	Original ACL
+ * \param[in]	new_ace	ACE to be added
+ * \param[out]	new_acl	Reallocated copy of the ACL with new item inserted
+ *
+ * \return	DER_SUCCESS or error code
+ */
+int
+daos_acl_add_ace_realloc(struct daos_acl *acl, struct daos_ace *new_ace,
+		struct daos_acl **new_acl);
+
+/**
+ * Remove an Access Control Entry from the list.
+ *
+ * This requires us to reconstruct and reallocate the ACL structure. Both the
+ * old structure and the new one must be freed by the caller. The old data is
+ * left unchanged.
+ *
+ * \param[in]	acl			Original ACL
+ * \param[in]	type			Principal type of the ACE to remove
+ * \param[in]	principal_name		Principal name of the ACE to remove
+ *					(NULL if type isn't user/group)
+ * \param[in]	principal_name_len	Length of the principal_name string
+ * \param[out]	new_acl			Reallocated copy of the ACL with the
+ *					ACE removed
+ *
+ * \return	DER_SUCCESS or error code
+ */
+int
+daos_acl_remove_ace_realloc(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal_name,
+		size_t principal_name_len, struct daos_acl **new_acl);
+
+/**
+ * Allocate a new Access Control Entry with an appropriately aligned principal
+ * name, if applicable.
+ *
+ * Only User and Group types use principal name.
+ *
+ * \param[in]	type			Type of principal for the ACE
+ * \param[in]	principal_name		Principal name will be added to the end
+ *					of the structure. For types that don't
+ *					use it, it is ignored. OK to pass NULL.
+ * \param[in]	principal_name_len	Length of the principal_name string
+ *
+ * \return	New ACE structure with an appropriately packed principal name,
+ *			length, and type set.
+ */
+struct daos_ace *
+daos_ace_alloc(enum daos_acl_principal_type type, const char *principal_name,
+		size_t principal_name_len);
+
+/**
+ * Free an Access Control Entry allocated by daos_ace_alloc().
+ *
+ * \param[in]	ace	ACE to be freed
+ */
+void
+daos_ace_free(struct daos_ace *ace);
+
+/**
+ * Get the length in bytes of an Access Control Entry.
+ * The entries have variable length.
+ *
+ * \param[in]	ace	ACE to get the size of
+ *
+ * \return	Size of ACE in bytes
+ */
+int
+daos_ace_get_size(struct daos_ace *ace);
```

Comment:
should be something like:
int daos_ace_get_size (struct daos_ace *ace, size_t *size);

Reply:
The actual size in this case is within the uint16_t space. Do we really need a separate param for this? I was thinking > 0 is the actual size, < 0 is an error.

Reply:
I went with Li Wei's suggestion above, and used ssize_t as the retval and updated the doxygen comment for clarity. If that's not acceptable, please let me know.